tests/tests/loopback.rs

```diff
@@ -0,0 +1,306 @@
+
+    /// Return the source underlay address.
+    fn underlay_ip(&self) -> std::net::Ipv6Addr {
+        self.cfg.phys_ip.into()
+    }
+}
+
+impl Drop for OptePort {
+    /// When this port is dropped, remove it from the underlying xde device.
+    fn drop(&mut self) {
+        let adm = match OpteAdm::open(OpteAdm::XDE_CTL) {
+            Ok(adm) => adm,
+            Err(e) => {
+                eprintln!("failed to open xde device on drop: {}", e);
+                return;
+            }
+        };
+        if let Err(e) = adm.delete_xde(&self.name) {
+            eprintln!("failed to delete xde on drop: {}", e);
+        }
+    }
+}
+
+/// This is resource handle for an xde device. It provides a few convenience
+/// methods for setting up global OPTE properties. It also removes the xde
+/// driver from the kernel when dropped. This is helpful for cleaning things up
+/// after a test run.
+struct Xde {}
+
+impl Xde {
+    /// Set the underlay data links that all OPTE ports will use.
+    fn set_xde_underlay(dev0: &str, dev1: &str) -> Result<()> {
+        let adm = OpteAdm::open(OpteAdm::XDE_CTL)?;
+        adm.set_xde_underlay(dev0, dev1)?;
+        Ok(())
+    }
+
+    /// Set the virtual to physical port mappings that all OPTE ports will use.
+    fn set_v2p(vip: &str, ether: &str, ip: &str) -> Result<()> {
+        let adm = OpteAdm::open(OpteAdm::XDE_CTL)?;
+        adm.set_v2p(&SetVirt2PhysReq {
+            vip: vip.parse().unwrap(),
+            phys: PhysNet {
+                ether: ether.parse().unwrap(),
+                ip: ip.parse().unwrap(),
+                vni: Vni::new(1701u32).unwrap(),
+            },
+        })?;
+        Ok(())
+    }
+}
+impl Drop for Xde {
+    /// When this object is dropped, remove the xde kernel module from the
+    /// underlying system.
+    fn drop(&mut self) {
+        let mut cmd = Command::new("pfexec");
+        cmd.args(["rem_drv", "xde"]);
+        if let Err(e) = cmd.output() {
+            eprintln!("failed to remove xde driver: {}", e);
+        }
+    }
+}
+
+/// This is an xde loopback topology. There are two zones, each with a vnic
+/// sitting atop an opte device. The opte device uses a pair of simnet
+/// devices as underlay links. These simnet devices are connected to each
+/// other.
+///
+///         zone a
+///     #============#
+///     | *--------* |    *-------*
+///     | | vopte0 |------| opte0 |         *------*
+///     | *--------* |    *-------*      +--| sim0 |--+
+///     #============#       |  *-----*  |  *------*  |
+///                          +--| xde |--+            |
+///     #============#       |  *-----*  |  *------*  |
+///     | *--------* |    *-------*      +--| sim1 |--+
+///     | | vopte1 |------| opte1 |         *------*
+///     | *--------* |    *-------*
+///     #============#
+///         zone b
+///
+/// The following system of overlay/underlay routes is set up
+///
+/// 10.0.0.1 -> fd47::/74 via sim1
+/// 10.0.0.2 -> fd74::/74 via sim0
```

Comment:
This is kind of a nit, but these addresses are pretty tricky for folks with dyslexia. It'd make it much harder to mess up with something like:

```
10.0.0.6 -> fd66::/66 via sim1
10.0.0.7 -> fd77::/77 via sim0
```

Those are arbitrary, but not "anagrams".

Reply:
Done